libyul/optimiser/UnusedFunctionReturnParameterPruner.cpp

```diff
@@ -0,0 +1,270 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libyul/optimiser/UnusedFunctionReturnParameterPruner.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <libyul/AsmDataForward.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+
+#include <libsolutil/CommonData.h>
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+using namespace solidity::yul::unusedFunctionsCommon;
+
+namespace
+{
+
+/**
+ * A helper class that removes expressions of the form `pop(a)` where a is an identifier. Generally,
+ * pop can be the discard function.
+ */
+class PopRemover: public ASTModifier
```

Comment:
Why can't we use the UnusedPruner for that?

Reply:
The more general step would be "remove ExpressionStatements that are movable".

Reply:
> Why can't we use the UnusedPruner for that?

Because of this: https://github.com/ethereum/solidity/pull/10080#discussion_r511770655

Reply:
It acutally does remove `pop(a)`, but it also removes `let a` in `let a := f()`. Since the result of `f` has to be used somehow, it is turned into `pop(f())`.